File: src/backend/codecc/common/common-util/src/main/java/com/tencent/devops/common/util/PathUtils.java
/*
 * Tencent is pleased to support the open source community by making BK-CODECC 蓝鲸代码检查平台 available.
 *
 * Copyright (C) 2019 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * BK-CODECC 蓝鲸代码检查平台 is licensed under the MIT license.
 *
 * A copy of the MIT License is included in this file.
 *
 *
 * Terms of the MIT License:
 * ---------------------------------------------------
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
 * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.tencent.devops.common.util;


import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class PathUtils
{
    private static Logger logger = LoggerFactory.getLogger(PathUtils.class);

    private static final int MIN_LENGTH = 6;
    public static final int DIR = 1;
    public static final int FILE = 2;

    public static void convertPaths(List<String> paths, ArrayList<String> storeArray, int type)
    {
        if (storeArray == null || paths == null)
        {
            return;
        }

        String postfix = "";
        if (type == DIR)
        {
            postfix = ".*";
        }

        for (String path : paths)
        {
            int wordCount = getWordCount(path, "/");
            int startWord = 0;
            int startIndex = 0;
            if (wordCount >= MIN_LENGTH)
            {
                startWord = wordCount - (wordCount * 2 / 3) + 1;
                char[] words = path.toCharArray();
                startIndex = getStartIndex(startWord, words, '/');
            }
            String regexPath = ".*" + path.substring(startIndex) + postfix;
            if (regexPath.contains("("))
            {
                regexPath = regexPath.replace("(", "\\(").replace(")", "\\)");
            }
            logger.debug("will add new ignore path:%s", regexPath);
            storeArray.add(regexPath);
        }
    }

    private static int getWordCount(String string, String splitChar)
    {
        return string.split(splitChar).length;
    }

    private static int getStartIndex(int startWord, char[] words, char splitChar)
    {
        if (words == null || words.length <= 0)
        {
            return 0;
        }

        int count = 0;
        for (int i = 0; i < words.length; i++)
        {
            if (words[i] == splitChar)
            {
                if (++count == startWord)
                {
                    return i;
                }
            }
        }
        return 0;
    }

    /**
     * 将文件的绝对路径转换成相对路径
     *
     * @param streamName
     * @param toolName
     * @param defectPaths
     * @return
     */
    public static Set<String> convertAbsolutePath2RelativePath(String streamName, String toolName, Set<String> defectPaths)
    {
        Set<String> relativePaths = new HashSet<String>(defectPaths.size());

        String splitRegex = streamName + "_" + toolName.toLowerCase();

        for (String absolutePath : defectPaths)
        {
            int i = absolutePath.lastIndexOf(splitRegex) + splitRegex.length();

            String relativePath = absolutePath.substring(i);
            relativePaths.add(relativePath);
        }

        return relativePaths;
    }

    /**
     * 将文件的绝对路径转换成相对路径
     *
     * @param streamName
     * @param toolName
     * @return
     */
    public static String convertAbsolutePath2RelativePath(String streamName, String toolName, String defectPath)
    {
        String splitRegex = streamName + "_" + toolName.toLowerCase();
        int i = defectPath.lastIndexOf(splitRegex) + splitRegex.length();

        String relativePath = defectPath.substring(i, defectPath.length());

        return relativePath;
    }

    /**
     * 根据工具侧上报的url和rel_path，截取url得到前面的值，然后拿最后一个/下的值
     *
     * @param url
     * @param relPath
     * @return
     */
    public static String getRelativePath(String url, String relPath)
    {
        if (org.apache.commons.lang.StringUtils.isEmpty(url) || org.apache.commons.lang.StringUtils.isEmpty(relPath))
        {
            return "";
        }

        if (!relPath.startsWith("/"))
        {
            relPath = "/" + relPath;
        }

        int index = url.lastIndexOf(relPath);
        if (index < 0)
        {
            return "";
        }
        //String tmpPath = url.substring(0, index);
        //return tmpPath.substring(tmpPath.lastIndexOf("/")) + relPath;

        return relPath;
    }

    /**
     * 检测路径是否匹配某个过滤路径
     *
     * @param path
     * @param ignorePaths
     * @return
     */
    public static boolean checkIfMaskByPath(String path, List<String> ignorePaths)
    {
        for (String regrex : ignorePaths)
        {
            if (path.matches(regrex))
            {
                return true;
            }
        }
        return false;
    }
}

File: src/backend/ci/ext/tencent/artifactory/biz-artifactory-tencent/src/main/kotlin/com/tencent/devops/artifactory/util/PathUtils.kt
/*
 * Tencent is pleased to support the open source community by making BK-CI 蓝鲸持续集成平台 available.
 *
 * Copyright (C) 2019 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * BK-CI 蓝鲸持续集成平台 is licensed under the MIT license.
 *
 * A copy of the MIT License is included in this file.
 *
 *
 * Terms of the MIT License:
 * ---------------------------------------------------
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 * the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT
 * LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
 * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.tencent.devops.artifactory.util

import com.tencent.devops.artifactory.pojo.FileInfo
import com.tencent.devops.common.api.constant.I18NConstant.BK_ILLEGAL_PATH
import com.tencent.devops.common.service.utils.HomeHostUtil
import com.tencent.devops.common.web.utils.I18nUtil
import java.net.URLEncoder
import java.nio.file.Paths
import java.util.regex.Pattern
import javax.ws.rs.BadRequestException

object PathUtils {
    fun checkAndNormalizeAbsPath(path: String): String {
        val normalizePath = Paths.get(path).normalize().toString()
        if (!normalizePath.startsWith("/")) {
            throw BadRequestException(I18nUtil.getCodeLanMessage(
                messageCode = BK_ILLEGAL_PATH
            ))
        }
        return normalizePath
    }

    fun normalize(relativePath: String): String {
        return Paths.get(relativePath).normalize().toString()
    }

    fun buildArchiveLink(projectId: String, pipelineId: String, buildId: String): String {
        return "${HomeHostUtil.outerServerHost()}/app/download/devops_app_forward.html?flag=buildArchive" +
                "&projectId=$projectId&pipelineId=$pipelineId&buildId=$buildId"
    }

    fun buildDetailLink(projectId: String, artifactoryType: String, path: String): String {
        return "${HomeHostUtil.outerServerHost()}/share/artifactoryDetail/?flag=artifactoryDetail" +
                "&projectId=$projectId&artifactoryType=$artifactoryType" +
                "&x-devops-project-id=$projectId&artifactoryPath=${
                    URLEncoder.encode(
                        path,
                        "utf-8"
                    )
                }"
    }

    fun isFolder(path: String): Boolean {
        return path.endsWith("/")
    }

    fun getParentFolder(path: String): String {
        return if (isFolder(path)) path.removeSuffix("${getFileName(path)}/")
        else path.removeSuffix(getFileName(path))
    }

    fun getFileName(path: String): String {
        return path.removeSuffix("/").split("/").last()
    }

    /**
     * 文件排序
     * 文件夹在文件之上
     * asc = true 文件名短在文件名长之上
     * asc = false 文件名短在文件名长之下
     */
    fun sort(fileInfoList: List<FileInfo>, asc: Boolean = true): List<FileInfo> {
        val ascInt = if (asc) 1 else -1
        return fileInfoList.sortedWith(Comparator { file1, file2 ->
            when {
                // 文件夹排在文件之上
                file1.folder && !file2.folder -> -1
                !file1.folder && file2.folder -> 1

                // 文件名短在文件名长之上
//                file1.name.length < file2.name.length -> -ascInt
//                file1.name.length > file2.name.length -> ascInt

                // 根据最后修改时间倒叙
                file1.modifiedTime < file2.modifiedTime -> ascInt
                file1.modifiedTime > file2.modifiedTime -> -ascInt

                // 类型相同长度相同，字母序排列
                else -> {
                    file1.name.compareTo(file2.name) * ascInt
                }
            }
        })
    }

    /**
     * 根据manifest文件路径解析镜像名称和版本
     * eg: /jq/manifest/1.0.0/manifest.json
     */
    fun getImageNameAndVersion(manifestPath: String): Pair<String, String> {
        val pattern = Pattern.compile("/(.*)/manifest/(.*)/manifest.json")
        val matcher = pattern.matcher(manifestPath)
        if (!matcher.find()) {
            throw IllegalArgumentException("illegal manifestPath: $manifestPath")
        }
        return Pair(matcher.group(1), matcher.group(2))
    }
}
